internal/inventory/gcpfetcher/fetcher_assets.go

```diff
@@ -0,0 +1,100 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package gcpfetcher
+
+import (
+	"context"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/cloudbeat/internal/inventory"
+	gcpinventory "github.com/elastic/cloudbeat/internal/resources/providers/gcplib/inventory"
+)
+
+type (
+	assetsInventory struct {
+		logger   *logp.Logger
+		provider inventoryProvider
+	}
+	inventoryProvider interface {
+		ListAllAssetTypesByName(ctx context.Context, assets []string) ([]*gcpinventory.ExtendedGcpAsset, error)
+	}
+	ResourcesClassification struct {
+		assetType      string
+		classification inventory.AssetClassification
+	}
+)
+
+var ResourcesToFetch = []ResourcesClassification{
+	{gcpinventory.CrmOrgAssetType, inventory.AssetClassificationGcpOrganization},
+	{gcpinventory.CrmFolderAssetType, inventory.AssetClassificationGcpFolder},
+	{gcpinventory.CrmProjectAssetType, inventory.AssetClassificationGcpProject},
+	{gcpinventory.ComputeInstanceAssetType, inventory.AssetClassificationGcpInstance},
+	{gcpinventory.ComputeFirewallAssetType, inventory.AssetClassificationGcpFirewall},
+	{gcpinventory.StorageBucketAssetType, inventory.AssetClassificationGcpBucket},
+	{gcpinventory.ComputeSubnetworkAssetType, inventory.AssetClassificationGcpSubnet},
+	{gcpinventory.IamServiceAccountAssetType, inventory.AssetClassificationGcpServiceAccount},
+	{gcpinventory.IamServiceAccountKeyAssetType, inventory.AssetClassificationGcpServiceAccountKey},
+}
+
+func newAssetsInventoryFetcher(logger *logp.Logger, provider inventoryProvider) inventory.AssetFetcher {
+	return &assetsInventory{
+		logger:   logger,
+		provider: provider,
+	}
+}
+
+func (f *assetsInventory) Fetch(ctx context.Context, assetChan chan<- inventory.AssetEvent) {
+	for _, r := range ResourcesToFetch {
+		f.fetch(ctx, assetChan, r.assetType, r.classification)
```

Comment:
we might want to run this in a goroutine with a wait group. thoughts?

Reply:
Good idea. The only limitation that comes to mind: is the rate-limit back-off safe to use in goroutines? If we hit API rate limits, would it work consistently for code in parallel?